File: src/imgproc.jl
import Base.convert
function split(image::Mat)
    hnd = ccall( (:split, "../libcv2"), Ptr{Ptr{Void}}, (Ptr{Void},), image.handle)
    images = pointer_to_array(hnd, 3)
    imageR = convert(Array{Mat, 1}, images)
    return imageR
end

function convert(::Type{Array{Mat, 1}}, ptrArray::Array{Ptr{Void}, 1})
    matArray = Mat[]
    for ptr in ptrArray
        append!(matArray, [Mat(ptr)])
    end
    return matArray
end

function columnStack(x::Array{Int, 1}, y::Array{Int, 1})
    pts = cv2.Point[]
    for i = 1:length(x)
        pt = cv2.Point(x[i], y[i])
        append!(pts, [pt])
    end
    return pts
end

function normarlizeT(array::Array{Int32, 1})
        big = array[1]
        small = array[1]

    for i = 2:length(array)
        temp = array[i]
        #println(temp)
        if temp > big
            big = temp
        end
        if temp < small
            small = temp
        end
    end

    rArray = Int[]

    for i = 1:length(array)
        append!(rArray, [convert(Int32, round(((array[i] - small)/(big - small))
        * 255))])
    end

    return rArray
end



function calcHist(images::Array{Mat,},
                  channels::Array{Int, 1},
                  #Null, #mask
                  histSize::Array{Int, 1},
                  ranges::Array{Float64, 1},
                  uniform::Bool = true,
                  accumulate::Bool = false)
    nimages = length(images)
    mask = Mat()

    imagesPtr = convertMatArray(images)
    hnd = ccall( (:calcHist, "../libcv2"),
              Ptr{Int32},
              (Ptr{Ptr{Void}}, Int, Ptr{Int}, Ptr{Void}, Int, Ptr{Int}, Ptr{Float64}, Bool, Bool),
              imagesPtr, nimages, channels, mask.handle, dims(images[1]), histSize, ranges, uniform, accumulate)
    hist = pointer_to_array(hnd, histSize[1])

    return hist
end

type Point
    x::Int
    y::Int
end

type cvPoint
    handle::Ptr{Void}
end

function cvPoint(x::Int, y::Int)
    pt = cvPoint(ccall( (:setPoint, "../libcv2"), Ptr{Void}, (Int, Int), x, y))
    return pt
end

function blur(image::Mat,
              size::(Int, Int),
              anchor::cv2.Point=cv2.Point(-1, -1),
              borderType::Int=0)
    sizeX = convert(Int32, size[1])
    sizeY = convert(Int32, size[2])
    sizePtr = [sizeX, sizeY]

    anchorPtr = cvPoint(anchor.x, anchor.y)
    result = Mat(ccall( (:blur, "../libcv2"), Ptr{Void}, (Ptr{Void}, Ptr{Int32},
    Ptr{Void}, Int), image.handle, sizePtr, anchorPtr.handle, borderType))

    return result
end

function polylines(image::Mat,
                   points::Array{cv2.Point, 1},
                   isClosed::Bool,
                   color::(Int, Int, Int),
                   thickness::Int = 1,
                   lineType::Int = 8,
                   shift::Int = 0)

    cvPts = convertPointArray(points)
    numOfPt = length(points) 
    colorPtr = convertColortoPtr(color)

    ccall( (:polylines, "../libcv2"), Void, (Ptr{Void}, Ptr{Ptr{Void}}, Int,
    Bool, Ptr{Int32}, Int, Int, Int), image.handle, cvPts, numOfPt, isClosed, colorPtr, thickness, lineType, shift)
end

function convertColortoPtr(color::(Int, Int, Int))
    r = convert(Int32, color[1])
    g = convert(Int32, color[2])
    b = convert(Int32, color[3])
    colorPtr = [r, g, b]
    return colorPtr
end

function convertMatArray(images::Array{cv2.Mat, 1})
    cvMat = Ptr{Void}[]
    for i in 1:length(images)
        append!(cvMat, [images[i].handle])
    end
    return cvMat
end

function convertPointArray(points::Array{cv2.Point, 1})
    cvPts = Ptr{Void}[]
    for i in 1:length(points)
        pt = cvPoint(points[i].x, points[i].y)
        append!(cvPts, [pt.handle])
    end
    return cvPts
end

File: src/cpp/imgproc.cpp
#include "imgproc.h"
#include <iostream>

// Note: I put some function that not belong the imgproc module.
// This is just because I wrap those function in same time.
// TODO: Reconstruct the file, correct the module function

using namespace std;

cv::Point* setPoint(int x, int y)
{
    cv::Point *pt = new cv::Point(x, y);
    return pt;
}

cv::Mat** split(const cv::Mat *src)
{
    cv::Mat** imageArray = new cv::Mat*[src->channels()];
    cv::Mat imageArrayTemp[src->channels()];
    cv::split(*src, imageArrayTemp);

    for (int i = 0; i < src->channels(); i++)
    {
        imageArray[i] = new cv::Mat;
        imageArrayTemp[i].copyTo(*(imageArray[i]));
    }

    return imageArray;
}

cv::Mat* blur(cv::Mat *img, int *size, cv::Point *anchor, int borderType)
{
    cv::Mat *result = new cv::Mat();
    cv::blur(*img, *result, cv::Size(5, 5), *anchor, borderType);

    return result;
}

void polylines(cv::Mat *img, cv::Point **pts, int npts, bool isClosed, int *color, int thickness, int lineType, int shift)
{
    using namespace cv;
    vector<Point> contour;
    for (int i = 0; i < npts; i++)
        contour.push_back(*pts[i]);

    // create a pointer to the data as an array of points (via a conversion to 
    // a Mat() object)
    const cv::Point *ptsx = (const cv::Point*) Mat(contour).data;

    cv::polylines(*img, &ptsx, &npts, 1, isClosed, Scalar(color[0],  color[1], color[2]), thickness, lineType, shift);
}

int* calcHist(const cv::Mat** images, int nimages, const int* channels, cv::Mat *mask, int dims, const int* histSize, const double* ranges, bool uniform, bool accumulate)
{
    cv::Mat *hist = new cv::Mat();

    float* histRanges = new float[nimages];

    // multiple to 2 is because every image range has two value
    for (int i = 0; i < nimages * 2; i++)
        histRanges[i] = (float)ranges[i];

    const float* const_histRanges = histRanges;

    using namespace cv;
  
     // Compute the histograms:
    calcHist( *images, nimages, channels, *mask, *hist, 1, histSize, &const_histRanges, uniform, accumulate);  

    // Copy the histogram to an int array
    // TODO: use int as arrary's type is just for test, it will correct later

    int *histArray = new int[*histSize];
    float* data = hist->ptr<float>(0);
    for ( int i = 0; i < *histSize; i++)
        histArray[i] = *data++;
  
     return histArray;
}
